### ISSUE TITLE
OneToOne relationship results in "Cyclic dependency" error

### ISSUE THREAD

**Issue by IRCraziestTaxi** [typeorm/typeorm#4385]
**Issue type:**

[ ] question
[ X ] bug report
[ ] feature request
[ ] documentation issue

**Database system/driver:**

Confirmed to occur in [TypeORM code](https://github.com/typeorm/typeorm/blob/master/src/persistence/SubjectTopoligicalSorter.ts) during entity validation I think? But to answer the question, have tried both:

[ X ] `mssql`
[ X ] `postgres`

**TypeORM version:**

0.2.18 (latest at time of opening this issue)

**Steps to reproduce or a small repository showing the problem:**

When using `@OneToOne` with cascade inserts, I have found that I receive a `Cyclic dependency` error. The error is thrown in [this file](https://github.com/typeorm/typeorm/blob/master/src/persistence/SubjectTopoligicalSorter.ts).

The problem does not occur with `@OneToMany` and `@ManyToOne` relationships.

Here is a small repro that I created:

https://github.com/IRCraziestTaxi/typeorm-one-to-one-cyclic-dependency

Note: Change `ormconfig.json` to use which ever DB driver, credentials, etc. you want to use and `npm install` the correct driver. I omitted it from `package.json` in order to avoid unnecessary bloat.

The migration is already checked in and a convenience script exists in `package.json`, so after modifying `ormconfig.json` as needed and installing the correct driver, run `npm run run-migrations` to apply the migration.

At a glance, here is the problematic relationship:

```ts
import { nameof } from "ts-simple-nameof";
import { Column, Entity, JoinColumn, OneToOne, PrimaryGeneratedColumn } from "typeorm";
import { UserProfile } from "./UserProfile";

@Entity()
export class User {
    @PrimaryGeneratedColumn()
    public id: number;

    @OneToOne(() => UserProfile, { cascade: true, onDelete: "CASCADE" })
    @JoinColumn({ name: nameof<User>(u => u.profileId) })
    public profile: UserProfile;

    @Column({ nullable: false })
    public profileId: number;

    @Column({ length: 100, nullable: false, unique: true })
    public username: string;
}
```

```ts
import { nameof } from "ts-simple-nameof";
import { Column, Entity, JoinColumn, OneToOne, PrimaryGeneratedColumn } from "typeorm";
import { User } from "./User";

@Entity()
export class UserProfile {
    @Column({ length: 100, nullable: false, unique: true })
    public email: string;

    @PrimaryGeneratedColumn()
    public id: number;

    @Column({ length: 100, nullable: false })
    public name: string;

    @OneToOne(() => User)
    @JoinColumn({ name: nameof<UserProfile>(p => p.userId) })
    public user: User;

    @Column({ nullable: false })
    public userId: number;
}
```

And here is the code attempting to perform the cascade insert:

```ts
import { LinqRepository } from "typeorm-linq-repository";
import { User } from "./entities/User";
import { UserProfile } from "./entities/UserProfile";

// ...

const profile = new UserProfile();
profile.email = `${Date.now()}@ex.com`;
profile.name = "John Doe";

const user = new User();
user.profile = profile;
user.username = `user${Date.now()}`;

const createdUser = await new LinqRepository(User).create(user);

// Error: Cyclic dependency: "UserProfile"
```

**EDIT**: Edited on 2019/7/5 for a little more clarity. Added migration instruction, entity creation code, and exact error message.

**Comment by IRCraziestTaxi** [comment]
Update: After running into problems with a similar relationship using EF Core (where each side of the one-to-one relationship tried to reference the other), I modified this relationship as follows:

```ts
import { Column, Entity, OneToOne, PrimaryGeneratedColumn } from "typeorm";
import { UserProfile } from "./UserProfile";

@Entity()
export class User {
    @PrimaryGeneratedColumn()
    public id: number;

    @OneToOne(() => UserProfile, { cascade: true, onDelete: "CASCADE" })
    public profile: UserProfile;

    // Removed profileId

    @Column({ length: 100, nullable: false, unique: true })
    public username: string;
}
```

```ts
import { nameof } from "ts-simple-nameof";
import { Column, Entity, JoinColumn, OneToOne, PrimaryColumn } from "typeorm";
import { User } from "./User";

@Entity()
export class UserProfile {
    @Column({ length: 100, nullable: false, unique: true })
    public email: string;

    @Column({ length: 100, nullable: false })
    public name: string;

    @OneToOne(() => User)
    @JoinColumn({
        name: nameof<UserProfile>(p => p.userId),
        // Added referencedColumnName
        referencedColumnName: nameof<User>(u => u.id)
    })
    public user: User;

    // Changed @Column to @PrimaryColumn
    @PrimaryColumn({ nullable: false })
    public userId: number;
}
```

So now the profile shares its key with User (if I'm understanding those decorators correctly).

However, now I get the following error:

```
Error: null value in column "userId" violates not-null constraint
```

I thought `cascade: true` on the `@OneToOne` decorator on `User.profile` would take care of setting the IDs on insert, but that does not appear to be the case.

**Comment by IRCraziestTaxi** [comment]
Another update: I also tried removing the `@JoinColumn` decorator from the `user` property on `UserProfile` and instead moved it to the `profile` property on `User`, which results in the same `Error: null value in column "userId" violates not-null constraint` error. I also tried having `@JoinColumn` on both relationships (while leaving the key on `UserProfile` as `@PrimaryColumn` rather than `@PrimaryGeneratedColumn` as it was before), but that reverted back to `Error: Cyclic dependency: "UserProfile"`.

**Comment by IRCraziestTaxi** [comment]
Final update:

I have determined that the initial `Cyclic dependency` error I received was indeed a valid error because I was misusing the `@JoinColumn` decorator by having it on both sides of the relationship.

I found #2758, which helped me straighten out what the relationship is supposed to look like (I have updated the sample repo) and confirms that the other error (`Error: null value in column "userId" violates not-null constraint`) I received when experimenting with having `@JoinColumn` on only one side of the relationship is indeed a bug.

Therefore, I will close this issue and track the referenced one instead.
